rpc_client/src/client.rs

```diff
@@ -0,0 +1,330 @@
+    fn build_request(&mut self, url: &Url) -> WsResult<Request> {
+		match Request::from_url(url) {
+			Ok(mut r) => {
+				let timestamp = time::UNIX_EPOCH.elapsed().unwrap().as_secs();
+				let hashed = format!("{}:{}", self.auth_code, timestamp).sha3();
+				let proto = format!("{:?}_{}", hashed, timestamp);
+				r.add_protocol(&proto);
+				Ok(r)
+			},
+			Err(e) => Err(WsError::new(WsErrorKind::Internal, format!("{}", e))),
+		}
+    }
+    fn on_error(&mut self, err: WsError) {
+		match mem::replace(&mut self.complete, None) {
+			Some(c) => c.complete(Err(RpcError::WsError(err))),
+			None => println!("warning: unexpected error"),
+		}
+    }
+    fn on_open(&mut self, _: Handshake) -> WsResult<()> {
+		match mem::replace(&mut self.complete, None) {
+			Some(c) => c.complete(Ok(Rpc {
+				out: mem::replace(&mut self.out, None).unwrap(),
+				counter: AtomicUsize::new(0),
+				pending: self.pending.clone(),
+			})),
+			// Should not be reachable
+			None => (),
+		}
+		Ok(())
+	}
+    fn on_message(&mut self, msg: Message) -> WsResult<()> {
+		match parse_response(&msg.to_string()) {
+			(Some(id), response) => {
+				match self.pending.remove(id) {
+					Some(c) => c.complete(response),
+					None => println!("warning: unexpected id: {}", id),
+				}
+			}
+			(None, response) => println!("warning: error: {:?}, {}", response, msg.to_string()),
+		}
+		Ok(())
+    }
+}
+
+/// Keeping track of issued requests to be matched up with responses
+#[derive(Clone)]
+struct Pending(Arc<Mutex<BTreeMap<usize, Complete<Result<JsonValue, RpcError>>>>>);
+
+impl Pending {
+	fn new() -> Self {
+		Pending(Arc::new(Mutex::new(BTreeMap::new())))
+	}
+	fn insert(&mut self, k: usize, v: Complete<Result<JsonValue, RpcError>>) {
+		self.0.lock().unwrap().insert(k, v);
+	}
+	fn remove(&mut self, k: usize) -> Option<Complete<Result<JsonValue, RpcError>>> {
+		self.0.lock().unwrap().remove(&k)
+	}
+}
+
+fn get_authcode(path: &PathBuf) -> Result<String, RpcError> {
+	match File::open(path) {
+		Ok(fd) => match BufReader::new(fd).lines().next() {
+			Some(Ok(code)) => Ok(code),
+			_ => Err(RpcError::NoAuthCode),
+		},
+		Err(_) => Err(RpcError::NoAuthCode)
+	}
+}
+
+/// The handle to the connection
+pub struct Rpc {
+	out: Sender,
+	counter: AtomicUsize,
+	pending: Pending,
+}
+
+impl Rpc {
+	/// Blocking, returns a new initialized connection or RpcError
+	pub fn new(url: &str, authpath: &PathBuf) -> Result<Self, RpcError> {
+		let rpc = try!(Self::connect(url, authpath).map(|rpc| rpc).wait());
+		rpc
+	}
+	/// Non-blocking, returns a future
+	pub fn connect(url: &str, authpath: &PathBuf)
+			   -> BoxFuture<Result<Self, RpcError>, Canceled> {
+		let (c, p) = oneshot::<Result<Self, RpcError>>();
+		match get_authcode(authpath) {
+			Err(e) => return done(Ok(Err(e))).boxed(),
+			Ok(code) => {
+				let url = String::from(url);
+				thread::spawn(move || {
+					// mem:replace Option hack to move `c` out
+					// of the FnMut closure
+					let mut swap = Some(c);
```

Comment:
yeah, `Option::take` exists for this purpose and should be used here. What's the guarantee that only one connection will trigger this factory, though? It doesn't seem so to me.